File: js/fx.js
/* jshint esversion: 6, devel: true, browser: true */

// The radius of a firework
const FIREWORK_SIZE = 100;
// The size of each firework particle
const PARTICLE_SIZE = 10;
// THe duration of a firework in seconds
const FIREWORK_DURATION = 1;
// The number of particles in a firework
const NUM_FIREWORK_PARTICLES = 16;

let fireworkManager = [];

class Firework {
   constructor(x, y) {
      this.x = x;
      this.y = y;
      this.interpolation = 0;
      switch (Math.floor(Math.random() * 4)) {
         case 0:
            this.color = COLOR_RED;
            break;
         case 1:
            this.color = COLOR_ORANGE;
            break;
         case 2:
            this.color = COLOR_GREEN;
            break;
         case 3:
            this.color = COLOR_BLUE;
            break;
      }
      fireworkManager.push(this);
   }

   isGone() {
      return this.interpolation >= 1;
   }

   render(delta, ctx2d) {
      // interpolate cuz we're cool
      if (this.interpolation < 1) {
         this.interpolation += delta / FIREWORK_DURATION;
         if (this.interpolation > 1)
            this.interpolation = 1;
      }

      // get the quintic value of the animation interpolation
      let t = quintEaseOut(this.interpolation);
      // the calculated particle size (in browser space)
      let s = toBrowserH((1 - t * t * t) * PARTICLE_SIZE);
      // distance of a particle from the center of the firework
      let d = t * FIREWORK_SIZE;

      // draw all the particles yay
      ctx2d.fillStyle = intToRGBText(this.color);
      for (let i = 0; i < NUM_FIREWORK_PARTICLES; i++) {
         let angle = 2 * Math.PI * i / NUM_FIREWORK_PARTICLES;
         let particleX = this.x + Math.sin(angle) * d;
         let particleY = this.y + Math.cos(angle) * d;
         ctx2d.fillRect(toBrowserX(particleX - s / 2), toBrowserY(particleY - s / 2), s, s);
      }
   }
}

const renderSpecialEffects = (delta, ctx2d) => {
   // render fireworks
   fireworkManager.forEach(function (firework) {
      firework.render(delta, ctx2d);
      if (firework.isGone())
         fireworkManager.splice(fireworkManager.indexOf(firework), 1);
   });
}

File: src/fx.rs
use rand::prelude::*;

use crate::block::BlockColor;

/// The radius of a firework
const FIREWORK_SIZE: u32 = 100;

/// The size of each firework particle
const PARTICLE_SIZE: u32 = 10;

/// The duration of a firework in seconds
const FIREWORK_DURATION: u32 = 1.0;

/// The number of particles in a firework
const NUM_FIREWORK_PARTICLES: u32 = 16;

struct FireworkManager {
    fireworks: Vec<Firework>
}

impl FireworkManager {
    pub fn push_firework(&mut self, firework: Firework) {
        self.fireworks.push(firework);
    }
}

pub struct Firework {
    pos: (f32, f32),
    interpolation: f32, // 0..=1
    color: BlockColor,
}

impl Firework {
   pub fn new(x: f32, y: f32) -> Self {
       Self {
           pos: (x, y),
           interpolation: 0.0,
           color: BlockColor::random(),
       }
   }

   fn is_gone(&self) {
       self.interpolation >= 1
   }

   fn render<CanvasContext>(delta: f32, ctx_2d: CanvasContext) {
      // interpolate cuz we're cool
      if (self.interpolation < 1) {
         self.interpolation += delta / FIREWORK_DURATION;
         self.interpolation = self.interpolation.min(1);
      }

      // get_the quintic value of the animation interpolation
      let t = quintEaseOut(self.interpolation);
      // the calculated particle size (in browser space)
      let s = toBrowserH((1 - t * t * t) * PARTICLE_SIZE);
      // distance of a particle from the center of the firework
      let d = t * FIREWORK_SIZE;

      // draw all the particles yay
      ctx_2d.fillStyle = intToRGBText(self.color);
      for i in 0..NUM_FIREWORK_PARTICLES {
         let angle = 2 * Math.PI * i / NUM_FIREWORK_PARTICLES;
         let particle_x = self.x + Math.sin(angle) * d;
         let particle_y = self.y + Math.cos(angle) * d;
         ctx_2d.fillRect(to_browser_x(particle_x - s / 2), to_browser_y(particle_y - s / 2), s, s);
      }
   }
}

fn render_special_effects<CanvasContext>(delta: f32, ctx_2d: CanvasContext) {
   // render fireworks
   fireworkManager.forEach(|firework| {
      firework.render(delta, ctx_2d);
      if (firework.isGone()) {
         fireworkManager.splice(fireworkManager.indexOf(firework), 1);
      }
   });
}
